components/engine_rocks_helper/src/sst_recovery.rs

```diff
@@ -0,0 +1,280 @@
+        self.damaged_files.iter().any(|f| f.name == sst_path)
+    }
+
+    // Cleans up obsolete damaged files and panics if some files are not handled in time.
+    //
+    // Acquire meta lock.
+    fn check_damaged_files(&mut self) {
+        if self.damaged_files.is_empty() {
+            return;
+        }
+
+        let mut new_damaged_files = self.damaged_files.clone();
+        new_damaged_files.retain(|f| {
+            if f.start_time.elapsed() > self.max_hang_duration {
+                self.set_panic_mark_and_panic(
+                    &f.name,
+                    &format!("recovery job exceeded {:?}", self.max_hang_duration),
+                );
+            }
+            self.check_overlap_damaged_regions(f.clone())
+        });
+        TIKV_ROCKSDB_DAMAGED_FILES.set(new_damaged_files.len() as i64);
+        self.damaged_files = new_damaged_files;
+    }
+
+    // Check whether the StoreMeta contains the region range, if it contains,
+    // recorded fault region ids to report to PD and add file info into `damaged_files`.
+    //
+    // Acquire meta lock.
+    fn check_overlap_damaged_regions(&mut self, file: FileInfo) -> bool {
+        let mut meta = self.store_meta.lock().unwrap();
+
+        // Find overlapping region ids.
+        // Condition:
+        // end_key > file.smallestkey
+        // start_key <= file.largestkey
+        let mut ids = vec![];
+        for (_, id) in meta
+            .region_ranges
+            .range((Excluded(file.smallest_key.clone()), Unbounded::<Vec<u8>>))
+        {
+            let region = &meta.regions[id];
+            if keys::enc_start_key(region) <= file.largest_key {
+                ids.push(*id);
+            }
+        }
+
+        let overlap = !ids.is_empty();
+        if overlap {
+            warn!(
+                "detected damaged regions overlapping with corrupted file";
+                "region_ids" => ?&ids,
+                "file" => &file.name,
+                "smallest_key" => ?&file.smallest_key,
+                "largest_key" => ?&file.largest_key,
+            );
+
+            // damaged file context should be updated.
+            for id in ids {
+                meta.damaged_regions_id.insert(id);
+            }
+            if !self.exist_scheduling_regions(&file.name) {
+                self.damaged_files.push(file);
+                TIKV_ROCKSDB_DAMAGED_FILES.inc();
+            }
+        } else {
+            fail_point!("sst_recovery_before_delete_files");
+            // The sst file can be deleted safely and set `include_end` to `true` otherwise the
+            // file with the same largest key will be skipped.
+            // Here store meta lock should be held to prevent peers from being added back.
+            self.db
+                .delete_files_in_range(&file.smallest_key, &file.largest_key, true)
+                .unwrap();
+            if self.db.get(&file.smallest_key).unwrap().is_some() {
```

Comment:
Why? I thought `delete_files_in_range` won't delete data in memtable or L0.

Reply:
`test_sst_recovery_overlap_range_sst_exist` may exactly tested this case？